File: Java/src/_0019_Remove_Nth_Node_From_End_of_List/Solution.java
package _0019_Remove_Nth_Node_From_End_of_List;

public class Solution {
    // 时间复杂度：O(n)
    // 链表遍历了两次
    // 空间复杂度：O(1)
    public ListNode removeNthFromEnd(ListNode head, int n) {
        int count = 0;  // 存储链表元素个数
        int delNum = 0;  // 存储正数要删除的元素位置，head位置为1

        ListNode curNode = head;
        while (curNode != null) {
            count++;
            curNode = curNode.next;
        }

        delNum = count - n + 1;

        ListNode dummyHead = new ListNode(0);
        dummyHead.next = head;
        ListNode p = dummyHead;
        for (int i = 0; i < delNum - 1; i++)
            p = p.next;

        // p后面的元素需要删除
        p.next = p.next.next;

        return dummyHead.next;
    }

    public static void main(String[] args) {
        int n = 2;
        // 创建链表，获取链表头结点
        ListNode head = (new MyLinkedList(new int[]{1, 2, 3, 4, 5})).getHead();

        head = (new Solution()).removeNthFromEnd(head, n);
        MyLinkedList.printLinkedList(head);
    }
}

File: Python/_0019_Remove_Nth_Node_From_End_of_List/Solution.py
# coding=utf-8
from MyLinkedList import *


# 执行用时：40 ms, 在所有 Python3 提交中击败了74.83%的用户
# 内存消耗：14.9 MB, 在所有 Python3 提交中击败了15.47%的用户
class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        dummy = ListNode()
        dummy.next = head
        p = dummy; q = dummy

        for i in range(n + 1):
            q = q.next
        while q:
            p = p.next
            q = q.next

        p.next = p.next.next

        return dummy.next


if __name__ == "__main__":
    head = MyLinkedList().createLinkedList([1, 2, 3, 4, 5])
    MyLinkedList().printLinkedList(Solution().removeNthFromEnd(head, 2))  # 1->2->3->5->NULL
